pkg/operator/controller/configurable-route/controller.go

```diff
@@ -0,0 +1,402 @@
+		client.InNamespace(operatorcontroller.OpenShiftConfigNamespace),
+	}
+}
+
+func (r *reconciler) deleteOrphanedRoles(componentRoutes []aggregatedComponentRoute) error {
+	existingHashes := map[string]struct{}{}
+	for _, cr := range componentRoutes {
+		existingHashes[cr.Hash] = struct{}{}
+	}
+
+	roleBindingList := &rbacv1.RoleBindingList{}
+	r.cache.List(context.TODO(), roleBindingList, allComponentRouteResources()...)
+	for _, item := range roleBindingList.Items {
+		expectedHash, ok := item.GetLabels()[componentRouteHashAnnotation]
+		if !ok {
+			return fmt.Errorf("Unable to find expected componentRoute hash label")
+		}
+
+		if _, ok := existingHashes[expectedHash]; !ok {
+			log.Info("deleting roleBinding", "name", item.GetName(), "namespace", item.GetNamespace())
+			if err := r.client.Delete(context.TODO(), &item); err != nil && !errors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *reconciler) deleteOrphanedRoleBindings(componentRoutes []aggregatedComponentRoute) error {
+	existingHashes := map[string]struct{}{}
+	for _, cr := range componentRoutes {
+		existingHashes[cr.Hash] = struct{}{}
+	}
+
+	roleBindingList := &rbacv1.RoleBindingList{}
+	r.cache.List(context.TODO(), roleBindingList, allComponentRouteResources()...)
+	for _, item := range roleBindingList.Items {
+		expectedHash, ok := item.GetLabels()[componentRouteHashAnnotation]
+		if !ok {
+			return fmt.Errorf("Unable to find expected componentRoute hash label")
+		}
+
+		if _, ok := existingHashes[expectedHash]; !ok {
+			log.Info("deleting roleBinding", "name", item.GetName(), "namespace", item.GetNamespace())
+			if err := r.client.Delete(context.TODO(), &item); err != nil && !errors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *reconciler) ensureServiceCertKeyPairSecretRole(owner metav1.Object, componentRoute aggregatedComponentRoute) (string, error) {
+	role := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: componentRoute.Name + "-",
+			Namespace:    r.config.SecretNamespace,
+			Labels: map[string]string{
+				componentRouteHashAnnotation: componentRoute.Hash,
+			},
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				Verbs:         []string{"get", "list", "watch"},
+				APIGroups:     []string{""},
+				Resources:     []string{"secrets"},
+				ResourceNames: []string{componentRoute.ServingCertificateName},
+			},
+		},
+	}
+
+	roleList := &rbacv1.RoleList{}
+	if err := r.cache.List(context.TODO(), roleList, componentRouteResources(componentRoute)...); err != nil {
+		return "", err
+	}
+
+	if len(roleList.Items) == 0 {
+		if _, _, err := resourceapply.ApplyRole(r.kclient.RbacV1(), r.eventRecorder, role); err != nil {
+			return "", err
+		}
+		return role.GetName(), nil
+	}
+
+	for i, curRole := range roleList.Items {
+		if i == 0 {
+			continue
+		}
+		if err := r.client.Delete(context.TODO(), &curRole); err != nil && !errors.IsNotFound(err) {
```

Comment:
strange to see `ApplyRole` and this delete before. A create should be enough, no?

Reply:
Updated in the latest version of the PR.